airbyte-integrations/connectors/source-google-analytics-v4/source_google_analytics_v4/source.py

```diff
@@ -0,0 +1,491 @@
+        self.credentials_json = json.loads(config["credentials_json"])
+        self.client_email = self.credentials_json["client_email"]
+        self.scope = "https://www.googleapis.com/auth/analytics.readonly"
+
+        super().__init__(
+            token_refresh_endpoint="https://oauth2.googleapis.com/token",
+            client_secret=self.credentials_json["private_key"],
+            client_id=self.credentials_json["private_key_id"],
+            refresh_token=None,
+        )
+
+    def refresh_access_token(self) -> Tuple[str, int]:
+        """
+        Calling the Google OAuth 2.0 token endpoint. Used for authorizing signed JWT.
+        Returns tuple with access token and token's time-to-live
+        """
+        response_json = None
+        try:
+            response = requests.request(method="POST", url=self.token_refresh_endpoint, params=self.get_refresh_request_params())
+
+            response_json = response.json()
+            response.raise_for_status()
+        except requests.exceptions.RequestException as e:
+            if response_json and "error" in response_json:
+                raise Exception(
+                    "Error refreshing access token {}. Error: {}; Error details: {}; Exception: {}".format(
+                        response_json, response_json["error"], response_json["error_description"], e
+                    )
+                ) from e
+            raise Exception(f"Error refreshing access token: {e}") from e
+        else:
+            return response_json["access_token"], response_json["expires_in"]
+
+    def get_refresh_request_params(self) -> Mapping[str, any]:
+        """
+        Sign the JWT with RSA-256 using the private key found in service account JSON file.
+        """
+        token_lifetime = 3600  # token lifetime is 1 hour
+
+        issued_at = time.time()
+        expiration_time = issued_at + token_lifetime
+
+        payload = {
+            "iss": self.client_email,
+            "sub": self.client_email,
+            "scope": self.scope,
+            "aud": self.token_refresh_endpoint,
+            "iat": issued_at,
+            "exp": expiration_time,
+        }
+
+        headers = {"kid": self.client_id}
+
+        signed_jwt = jwt.encode(payload, self.client_secret, headers=headers, algorithm="RS256")
+
+        return {"grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer", "assertion": str(signed_jwt)}
+
+
+class SourceGoogleAnalyticsV4(AbstractSource):
+    """Google Analytics lets you analyze data about customer engagement with your website or application."""
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        try:
+            url = f"{GoogleAnalyticsV4TypesList.url_base}"
+
+            authenticator = GoogleAnalyticsOauth2Authenticator(config)
+
+            session = requests.get(url, headers=authenticator.get_auth_header())
+            session.raise_for_status()
+
+            custom_reports = config.get("custom_reports")
```

Comment:
let's add an additional try/except block for this part with custom json  with the descriptive message if the custom report is not a valid json.

There is big probability that user could load broken json

Reply:
Updated